### PR TITLE
Add concept of AuthSessions

### PR DESCRIPTION
**Pull request by Kegsay**
Auth sessions are a single auth process between a user and an auth realm. As
such, they are keyed off the tuple of `(user_id, realm_id)`.

Only the realm which they belong to knows how to construct them, hence all
"load" sections require an `AuthRealm` to be extracted first.

Currently I pass in a `json.RawMessage` rather than factory initialise and
clobber public fields based on the JSON, we can always change that if need be
later down the line.

Overall, this feels really nice (when starting to add in GH auth, everything I
wanted was already there in the right place waiting for me).

### CODE REVIEW

File: src/github.com/matrix-org/go-neb/database/schema.go [hunk 1]
```diff
@@ -268,3 +277,64 @@ func updateRealmTxn(txn *sql.Tx, now time.Time, realm types.AuthRealm) error {
...
+	var realmType string
+	var sessionJSON []byte
+	row := txn.QueryRow(selectAuthSessionSQL, realmID, userID)
+	if err := row.Scan(&realmType, &sessionJSON); err != nil {
+		return nil, err
+	}
+	realm := types.CreateAuthRealm(realmID, realmType)
+	if realm == nil {
+		return nil, fmt.Errorf("Cannot create realm of type %s", realmType)
+	}
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/go-neb/database/schema.go]: Might be worth plucking out the auth realm JSON from the database to fillout the common fields before calling AuthSession.

**Comment by NegativeMjark** [comment]
What's the intended lifetime of a AuthSession? Does it contain the token? or is it just temporary bit of state for getting a token?

File: src/github.com/matrix-org/go-neb/database/schema.go (hunk 1 above)
Comment (Kegsay) [inline src/github.com/matrix-org/go-neb/database/schema.go]: done

**Comment by Kegsay** [comment]
The intention is that the `AuthSession` will contain the token. If a session exists for a user ID / realm ID tuple, then either they have successfully authed or they are in the process of authing (only the realm knows).

**Comment by NegativeMjark** [comment]
Okay. Looks good enough to me to make an attempt at actually implementing the github auth. It'll be be clearer whether this API shape makes sense then.

**Pull request merged**